Add eBPF pros/cons blog post (#129)
pyroscope-io/docs: blog/ebpf-profiling-pros-cons.mdx

image: https://user-images.githubusercontent.com/23323466/192919567-161b3f4b-8998-4968-a399-ae9f9ca8d1b5.png
sidebar_label: Profiling with eBPF
slug: /ebpf-profiling-pros-cons
date: "2022-09-29"

authors:
  - name: Ryan Perry
    url: https://github.com/Rperry2174
    image_url: https://avatars.githubusercontent.com/u/23323466?v=4
---

<img width="100%" src="https://user-images.githubusercontent.com/23323466/192919567-161b3f4b-8998-4968-a399-ae9f9ca8d1b5.png" />

## What is eBPF?

At its root, eBPF takes advantage of the kernel’s privileged ability to oversee and control the entire system.
With eBPF you can run sandboxed programs in a privileged context such as the operating system kernel.
To better understand the implications and learn more check out this [blog post](https://ebpf.io/what-is-ebpf)
which goes into much more detail. For profiling this typically means running a program that pulls stacktraces
for the whole system at a constant rate (e.g 100Hz).

![image](https://user-images.githubusercontent.com/23323466/188525916-0e33c64a-9a7a-4fe2-84d6-91111c4d6f39.png)

As you can see in the diagram, some of the most popular use cases for eBPF are related to Networking, Security,

> and most relevant to this blog post -- observability (logs, metrics, traces, and profiles).

Reviewer: ```suggestion
and most relevant to this blog post — observability (logs, metrics, traces, and profiles).
```

is more frequently being considered a first-class citizen in observability suites. 

For example, OpenTelemetry has kicked off
efforts to [standardize profiling](https://github.com/open-telemetry/oteps/blob/main/text/profiles/0212-profiling-vision.md)
in order to enable more effective observability. For more information on those efforts check out
the [#otel-profiling](https://cloud-native.slack.com/archives/C03J794L0BV) channel on the CNCF slack!

## Current state of eBPF profiling

There's a wide spectrum of challenges with eBPF profiling that can drastically change the value you're able to receive from utilizing it.
This is a list roughly sorted from "easier" fairly standard functionality to "harder" more advanced functionality:

- Collecting data with low overhead
- Auto-tagging profiles with available metadata
- Finding the best _format_ in which to send profiling data
- Dealing with collecting, storing, and applying symbols
- Finding ways to deal with _lack_ of symbols
- Finding efficient ways to do other kinds of profiling such as memory profiling
- Finding ways to apply tags dynamically and/or arbitrarily

> This blog will focus on Pyroscope's unique approach to solving these issues.

Reviewer: I would maybe drop this whole section:
* The list of challenges is somewhat similar to list of pros and cons.
* We say that "This blog will focus on Pyroscope's unique approach to solving these issues." but we don't really do that

> ## Current state of eBPF profiling
> 
> There's a wide spectrum of challenges with eBPF profiling that can drastically change the value you're able to receive from utilizing it.
> This is a list roughly sorted from "easier" fairly standard functionality to "harder" more advanced functionality:
> 
> - Collecting data with low overhead
> - Auto-tagging profiles with available metadata
> - Finding the best _format_ in which to send profiling data
> - Dealing with collecting, storing, and applying symbols
> - Finding ways to deal with _lack_ of symbols
> - Finding efficient ways to do other kinds of profiling such as memory profiling
> - Finding ways to apply tags dynamically and/or arbitrarily
> 
> This blog will focus on Pyroscope's unique approach to solving these issues.

Reviewer: ```suggestion

```

I would maybe drop this whole section: (Current state of eBPF profiling)
* The list of challenges is somewhat similar to list of pros and cons.
* We say that "This blog will focus on Pyroscope's unique approach to solving these issues." but we don't really do that

  </TabItem>
</Tabs>

### Profiling interpreted languages (Ruby, Python, etc.)
With interpreted languages like Ruby or Python, stacktraces in their runtimes are not easily accessible from the kernel. 
As a result, the eBPF profiler is not able to parse user-space stack traces for interpreted languages. You can see how
the kernel interprets stack traces of compiled languages (go) vs how the kernel interprets stack traces from interpreted languages (ruby/python)
in the examples

<Tabs
  defaultValue="nativePython"
  values={[
    {label: 'Native Python', value: 'nativePython'},
    {label: 'eBPF Python', value: 'ebpfPython'},
    {label: 'Native Ruby', value: 'nativeRuby'},
    {label: 'eBPF Ruby', value: 'ebpfRuby'}
  ]}>
  <TabItem value="nativePython">
    <Flamegraph profileData={PythonNative}></Flamegraph>
  </TabItem>
  <TabItem value="ebpfPython">
    <Flamegraph profileData={PythonEbpf}></Flamegraph>
  </TabItem>
  <TabItem value="nativeRuby">
    <Flamegraph profileData={RubyNative}></Flamegraph>
  </TabItem>
  <TabItem value="ebpfRuby">
    <Flamegraph profileData={RubyEbpf}></Flamegraph>
  </TabItem>

</Tabs>

## How to use eBPF for cluster level profiling

> Using Pyroscope's auto-tagging feature in th eBPF integration you can get a breakdown of cpu usage by kubernetes metadata.

Reviewer: ```suggestion
Using Pyroscope's auto-tagging feature in the eBPF integration you can get a breakdown of cpu usage by kubernetes metadata.
```

In this case we can see which namespace is consuming the most cpu resources for our [demo instance](https://demo.pyroscope.io/explore?query=rideshare-cluster-ebpf.cpu%7B%7D&groupBy=k8s_pod_namespace&groupByValue=All)
after adding Pyroscope with two lines of code:

```shell
# Add Pyroscope eBPF integration to your kubernetes cluster
helm repo add pyroscope-io https://pyroscope-io.github.io/helm-chart
helm install pyroscope-ebpf pyroscope-io/pyroscope-ebpf
```

![image](https://user-images.githubusercontent.com/23323466/192881179-996930b2-5307-4803-ac9b-09e1bf0f97f2.png)

and you can also see the flamegraph representing CPU utilization for the entire cluster:
![image](https://user-images.githubusercontent.com/23323466/192880734-ab0a9503-cc7e-4c53-b677-6b2e8d0f690e.png)

> This is how we profile our applications internally:
> - eBPF for our kubernetes cluster
> - ruby gem, pip package, go client, and java client _with tags_ for our language microservices and testing suites
> - Lambda extension for our lambda functions

Reviewer: I would maybe spell this out more.

```suggestion
Internally, we use a variety of integrations to get both a high level overview of what's going on in our cluster, but also a very detailed view for each runtime that we use:
- We use our eBPF integration for our kubernetes cluster
- We use [ruby gem](https://pyroscope.io/docs/ruby/), [pip package](https://pyroscope.io/docs/python/), [go client](https://pyroscope.io/docs/golang/), and [java client](https://pyroscope.io/docs/java/) _with tags_ for our k8s services and github action test suites.
- We use our [lambda extension](https://pyroscope.io/docs/aws-lambda/) to profile the code inside lambda functions
```
